### PR TITLE
mgr/dashboard: User password should be optional

### PR DESCRIPTION
**Pull request by ricardoasmarques**
With this PR, user password will be optional which is a requirement to introduce SSO support.

If no password is defined for a user, local authentication will fail.

Fixes: https://tracker.ceph.com/issues/36031

Signed-off-by: Ricardo Marques <rimarques@suse.com>

### CODE REVIEW

**Comment by rjfd** [comment]
@ricardoasmarques is it possible to write a test in `test_auth` that tests the case where a user without a password defined cannot login?

File: src/pybind/mgr/dashboard/services/access_control.py [hunk 1]
```diff
@@ -669,9 +671,10 @@ def __init__(self):
     def authenticate(self, username, password):
         try:
             user = ACCESS_CTRL_DB.get_user(username)
-            pass_hash = password_hash(password, user.password)
-            if pass_hash == user.password:
-                return user.permissions_dict()
+            if user.password:
```
Comment (LenzGr) [inline src/pybind/mgr/dashboard/services/access_control.py]: What happens if the password is either incorrect or empty?
What error/message do we return here?

Comment (ricardoasmarques) [reply src/pybind/mgr/dashboard/services/access_control.py]: If no password is specified, user will see the same error that is displayed for incorrect password:

screenshot from 2018-09-18 13-30-30

**Comment by ricardoasmarques** [comment]
@rjfd I've added a test that tests the case where no password is defined.

**Comment by callithea** [comment]
QA run was successful: http://pulpito.ceph.com/laura-2018-09-18_22:07:59-rados:mgr-wip-lpaduano-testing-24128-distro-basic-smithi/

File: src/pybind/mgr/dashboard/services/access_control.py (hunk 1 above)
Comment (Devp00l) [reply src/pybind/mgr/dashboard/services/access_control.py]: The user won't get this message in the current UI as the login page requires a password to be typed in before you can click on login.

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-form/user-form.component.ts [hunk 2]
```diff
@@ -83,21 +83,10 @@ export class UserFormComponent implements OnInit {
     });
     if (this.mode === this.userFormMode.editing) {
       this.initEdit();
-    } else {
-      this.initAdd();
     }
   }
```
Comment (Devp00l) [inline src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-form/user-form.component.ts]: Could you please add a test for the minimum required filled input fields?

**Comment by ricardoasmarques** [comment]
> The user won't get this message in the current UI as the login page requires a password to be typed in before you can click on login.

If no password is defined in the user database, user will see this error message when trying to login with any password (empty or not). And yes, if using the UI, the frontend will force a non empty password but we have to keep in mind that we cannot assume that all requests come from the UI.

Just to clarify, is this comment a request to change anything in the PR?

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-form/user-form.component.ts (hunk 2 above)
Comment (ricardoasmarques) [reply src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-form/user-form.component.ts]: The only required field ATM is the `username`, and we already have a test for that field:

https://github.com/ceph/ceph/blob/bdc93a3550b9e2d4f4ad41cd818da713d94bfc43/src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-form/user-form.component.spec.ts#L88

What to you really want to test?

Comment (Devp00l) [reply src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-form/user-form.component.ts]: Just if the form is invalid if the username is not set, and that it is valid if it is set.

**Comment by ricardoasmarques** [comment]
@Devp00l I've addressed your comments

**Pull request merged**